### PR TITLE
add pickle functionality. fixes #473

### LINKED ISSUE #473: Can't pickle _tskit.TreeSequence objects?
Are `TreeSequence` objects meant to be pickleable? 
```
>>> import msprime as msp
>>> msp.__version__
'0.7.4'
>>> import pickle
>>> ts = msp.simulate(2)
>>> pickle.dumps(ts)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't pickle _tskit.TreeSequence objects
```
On the other hand, `ts.tables` can be pickled. So maybe this is intentional?

### PR DESCRIPTION
**Pull request by terhorst**
fairly straightforward. for lack of a better idea, I stuck the test at the end of `test_file.py`.

### CODE REVIEW

File: python/tests/test_file_format.py [hunk 1]
```diff
@@ -779,3 +780,20 @@ def test_load_bad_formats(self):
         with open(self.temp_file, "wb") as f:
             f.write(os.urandom(8192))
         self.assertRaises(exceptions.FileFormatError, tskit.load, self.temp_file)
+
+
+class TestPickle(TestFileFormat):
+    """
+    Test pickling of a TreeSequence.
+    """
```
Comment (jeromekelleher) [inline python/tests/test_file_format.py]: This is great! To generalise a little bit, I'd do this:
```python
def verify_round_trip(self, ts):
    pkl = pickle.dumps(ts)
    tsp = pickle.loads(pkl)
    self.assertEqual(ts.tables, tsp.tables)

def test_simple(self):
    self.verify_round_trip(msprime.simulate(10, random_seed=2))

def test_recombination(self):
    self.verify_round_trip(msprime.simulate(10, recombation_rate=1, random_seed=2))

def test_mutations(self):
    self.verify_round_trip(msprime.simulate(10, mutation_rate=1, random_seed=2))

def test_migrations(self):
    ts = msprime.simulate(
          population_configurations=[
              msprime.PopulationConfiguration(10),
              msprime.PopulationConfiguration(10)],
          migration_matrix=[[1,0], [0, 1]],
          record_migrations=True,
          random_seed=2)
         self.verify_round_trip(ts)
```
That basically catches all the important stuff.

**Comment by benjeffery** [comment]
Thanks @terhorst, nice addition. Looking at the tests maybe https://github.com/tskit-dev/tskit/blob/master/python/tests/test_highlevel.py#L577 would be a good class to add it to? Mainly thinking this as it is `TreeSequence` that the new methods are on.

We've just moved to `black` for python formatting so to fix the CI error you need to run `black` to fix the formatting. I'm trying to make this process easier in https://github.com/tskit-dev/tskit/pull/471 if you have any thoughts on that then we'd love to hear them.

**Comment by benjeffery** [comment]
Comment collison!

**Comment by jeromekelleher** [comment]
> Comment collison!

Consistent comments, though, which is good!

> Thanks @terhorst, nice addition. Looking at the tests maybe https://github.com/tskit-dev/tskit/blob/master/python/tests/test_highlevel.py#L577 would be a good class to add it to? Mainly thinking this as it is TreeSequence that the new methods are on.

I agree, we can move this whole test class (as I outlined above) into tests/test_highlevel.py and that's a good home for it.

**Comment by jeromekelleher** [comment]
Perfect, thanks @terhorst!

**Pull request merged**